### DISCUSSION TITLE
How to use PurgeCSS for HTML pages generated using a template engine

### DISCUSSION THREAD

**Discussion by vralle** [webdiscus/html-bundler-webpack-plugin#31] [Q&A] [answered]
Hi there!

I use:

- html-bundler-webpack-plugin
- Template engines such as ETA
- Bootstrap SCSS

I would like to remove unused styles for each HTML page compiled using a template engine

[PurgeCSS](https://purgecss.com/) can remove unused styles. I tried [purgecss-webpack-plugin](https://purgecss.com/plugins/webpack.html). But it can't see template parts, if I specify uncompiled templates in the arguments.
 
I don't have much experience using webpack hooks, and it would be great to get some advice on how best to implement style cleanup for each page created using template system.

My case.

error-template.eta
```html
<!DOCTYPE html>
<html lang="en" class="h-100">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>{{= title }}</title>
        <meta name="description" content="{{= description }}">
        <link rel="stylesheet" href="./scss/error.scss">
    </head>
    <body class="d-flex w-100 h-100 mx-auto flex-column">
        {{~ include("./parts/header.eta") }}
        {{~ include("./parts/main.eta") }}
        {{~ include("./parts/footer.eta") }}
    </body>
</html>
```

webpack config example:
```js
const sass = require('sass');
const HtmlBundlerPlugin = require('html-bundler-webpack-plugin');

// Configs
const isProduction = process.env.NODE_ENV === 'production';
const postCssConfig = require('./config/postcss.config');

// Paths
const srcDir = path.resolve(__dirname, 'src');
const outputDir = path.resolve(__dirname, 'public');
const outputAssetDir = 'static';

/** @type {import('webpack').Configuration} */
module.exports = {
    mode: isProduction ? 'production' : 'development',
    output: {
        path: outputDir,
        filename: `${outputAssetDir}/[name].[fullhash:7][ext][query]`,
        publicPath: '',
        assetModuleFilename: `${outputAssetDir}/[name][ext][query]`,
        clean: true,
    },
    plugins: [
        new HtmlBundlerPlugin({
            preprocessor: 'eta',
            preprocessorOptions: {
                tags: ['{{', '}}'],
            },
            entry: {
                index: `${srcDir}/index.eta`,
                'error-404': {
                    import: `${srcDir}/error-template.eta`,
                    data: {
                        title: 'Not found',
                        description: 'The page has been moved, deleted, or the link is incorrect. We apologize for the inconvenience.',
                    },
                },
                'error-500': {
                    import: `${srcDir}/error-template.eta`,
                    data: {
                        title: 'Server error',
                        description: 'The server cannot respond to the request. Please try again later. We apologize for the inconvenience.',
                    },
                },
                'error-503': {
                    import: `${srcDir}/error-template.eta`,
                    data: {
                        title: 'We are improving our website',
                        description: 'This page is currently under maintenance. We will back soon with new and improved features.'
                    },
                },
            },
            css: {
                filename: `${outputAssetDir}/css/[name].[contenthash:7].css[query]`,
                inline: true,
            },
        }),
    ],
    module: {
        rules: [
            {
                test: /\.s?css$/i,
                use: [
                    {
                        loader: 'css-loader',
                        options: {},
                    },
                    isProduction && {
                        loader: 'postcss-loader',
                        options: {},
                    },
                    {
                        loader: 'sass-loader',
                        /** @type {import('sass-loader').Options} */
                        options: {
                            api: 'modern',
                            webpackImporter: true,
                            warnRuleAsWarning: true,
                            implementation: sass,
                            /** @type { import('sass').Options } */
                            sassOptions: {
                                loadPaths: ['node_modules'],
                                outputStyle: 'expanded',
                            },
                        },
                    },
                ].filter(Boolean),
            },
            {
                test: /\.(jpe?g|png|webp|svg|ico)$/i,
                type: 'asset',
            },
        ],
    },
};

```
Any idea?

**Comment by webdiscus** [comment]
Hello @vralle,

Thanks for the interesting use case.
I'll try to reproduce your case and give feedback later.

**Comment by webdiscus** [comment]
@vralle 

I have created the [purgecss example](https://github.com/webdiscus/html-bundler-webpack-plugin/tree/master/examples/purgecss).
It works fine.

---
add bootstrap in your SCSS file `error.scss`
```scss
@use 'bootstrap';

// your styles
```

add PurgeCSS plugin in Webpack config:
```js
const path = require('path');
const glob = require('glob'); // <= install

const HtmlBundlerPlugin = require('html-bundler-webpack-plugin');
const { PurgeCSSPlugin } = require('purgecss-webpack-plugin');  // <= install

// Configs
const isProduction = process.env.NODE_ENV === 'production';
const postCssConfig = require('./config/postcss.config');

// Paths
const srcDir = path.resolve(__dirname, 'src');
const outputDir = path.resolve(__dirname, 'public');
const outputAssetDir = 'static';

/** @type {import('webpack').Configuration} */
module.exports = {
    mode: isProduction ? 'production' : 'development',
    output: {
        path: outputDir,
        publicPath: '',
        assetModuleFilename: `${outputAssetDir}/[name][ext][query]`,
        clean: true,
    },
    plugins: [
        new HtmlBundlerPlugin({
            preprocessor: 'eta',
            preprocessorOptions: {
                tags: ['{{', '}}'],
            },
            entry: {
                index: `${srcDir}/index.eta`,
                'error-404': {
                    import: `${srcDir}/error-template.eta`,
                    data: {
                        title: 'Not found',
                        description: 'The page has been moved, deleted, or the link is incorrect. We apologize for the inconvenience.',
                    },
                },
                'error-500': {
                    import: `${srcDir}/error-template.eta`,
                    data: {
                        title: 'Server error',
                        description: 'The server cannot respond to the request. Please try again later. We apologize for the inconvenience.',
                    },
                },
                'error-503': {
                    import: `${srcDir}/error-template.eta`,
                    data: {
                        title: 'We are improving our website',
                        description: 'This page is currently under maintenance. We will back soon with new and improved features.'
                    },
                },
            },
            js: {
                 // define JS output filename here to keep same settings in one place
                 filename: `${outputAssetDir}/js/[name].[contenthash:7].js[query]`, // <= important: use the `contenthash`
            },
            css: {
                filename: `${outputAssetDir}/css/[name].[contenthash:7].css[query]`,
                inline: true,
            },
        }),
        isProduction && new PurgeCSSPlugin({
          // scan *.html and *.eta files in src subdirs
          paths: glob.sync(`${srcDir}/**/*.+(html|eta)`, { nodir: true }),
        }),
    ].filter(Boolean),
    module: {
        rules: [
            {
                test: /\.s?css$/i,
                use: [
                    {
                        loader: 'css-loader',
                        options: {},
                    },
                    isProduction && {
                        loader: 'postcss-loader',
                        options: {},
                    },
                    {
                        loader: 'sass-loader',
                        /** @type {import('sass-loader').Options} */
                        options: {},
                    },
                ].filter(Boolean),
            },
            {
                test: /\.(jpe?g|png|webp|svg|ico)$/i,
                type: 'asset',
            },
        ],
    },
};
```

**Comment by vralle** [comment]
Hi @webdiscus. Thanks for the help!
I had the same experience. The example works.

But when I tried to make the styles for the error-* pages as small as possible, I couldn't. The reason is obvious, the styles of simple error-* pages contain the styles of the landing page index.* and others. For this reason, the size of error-* pages will increase significantly.

To keep the size of styles to a minimum, you need to clean each page individually. I imagined it looking like this:
```js
...
postprocess(content, info, compilation) {
  const compiledStyles = howGetCompiledStyleForPage(arguments);

  const cleanedStyles = await new PurgeCSS().purge({
    content: [
      /** @type {import('purgecss').RawContent | string} */
      {
        raw: content,
        extension: 'html',
      }
    ],
    /**  */
    css: [
      /** @type {import('purgecss').RawCSS | string} */
      compiledStyles
    ],
    fontFace: true,
    keyframes: true,
    variables: true,
  });

  howToSaveCleanedStyles(cleanedStyles);

  return content;
}
...
```
Here I was lost due to lack of understanding of webpack hooks 😕

At the moment I'm using  [postcss-purgecss](https://purgecss.com/plugins/postcss.html)

purgeCss.plugin.js:
```js
/**
 * @typedef {import('@fullhuman/postcss-purgecss').UserDefinedOptions} PostCssPurgeCssOptions
 * @typedef {PostCssPurgeCssOptions} Options
 */

const path = require('node:path');
const { globSync } = require('glob');

const paths = {
  src: path.resolve(__dirname, '../../src'),
  output: path.resolve(__dirname, '../../public'),
};

/**
 * @type {{string: Options}} plugin
 */
const plugin = {
  '@fullhuman/postcss-purgecss': {
    contentFunction: (sourceInputFileName) => {
      // ? Support for error pages
      if (/error\.scss$/.test(sourceInputFileName)) {
        return [`${paths.output}/error-404.html`];
      }

      return globSync(
        `${paths.output}/*.html`,
        {
          nodir: true,
          ignore: {
            ignored: (globPath) => /error-[\w\d-]+\.html$/.test(globPath.name),
          },
        },
      );
    },
    keyframes: true,
    variables: true,
  },
};

module.exports = plugin;
```

postcss.config.js:
```js
const purgeCss = require('./purgeCss.plugin.js');

/** @type {import('postcss-load-config').Config} */
const config = {
  map: false,
  plugins: {
    ...purgeCss,
    autoprefixer: {
      cascade: false,
    },
  },
};

module.exports = config;
```

This works as I planned. But it has an obvious drawback. I need to specify already pre-compiled html files, which causes errors when updating the template. And the code will be hard to maintain when the number of pages starts to increase 🥴

**Comment by webdiscus** [comment]
ah, now I understand the whole problem.

Hmm, this is realy a very complex use case. I don't have a ready answer.

The variant with `postprocess` will not work because:
- the `postprocess` called in the `sync` webpack hook. Async functions are here impossible.
- in the `postprocess` you can modify rendered HTML content only
- in the `postprocess` you can not get compiled style for page, because firstly is rendered self page and only at next iteration will be compiled CSS from the source style file parsed in HTML

I can imagine this as `cssPostprocess` for CSS when CSS is compiled, then would be available CSS with its issuer (HTML page). But, in this case fails two important features:
- the integrity will be wrong (recalculation is impossible  because the processing chain is mega complex)
- source-map will be wrong (recalculation is impossible  because the processing chain is mega complex)

_I have to think about this variant..._

Therefore, now the possible solution is your variant with postcss plugin. It's not comfortable, but it works.

P.S. do you know a solution using other plugins, e.g. `html-webpack-plugin`? If you find it, let me know and I'll try to implement it.

**Comment by webdiscus** [comment, marked as answer]
I think the complexity of implementation is  much higher than using `postcss-purgecss` with the `contentFunction()` therefore solve your special case  on the `postcss-purgecss` side.

I have created the [purgecss-postcss](https://github.com/webdiscus/html-bundler-webpack-plugin/tree/master/examples/purgecss-postcss) example where each page contains own style file and compiled CSS contains only used on the page classes.

Idea:
- bind style with html via EntryPoint `name`
- each output HTML file has the EntryPoint `name` (the key of the `entry` option)
- each page must have own style file with the filename containing `name`, e.g. for `about.html` -> `about.scss`
- use the `contentFunction` to retrieve the `name` form style filename
- the `contentFunction` returns the HTML file based in the `name` of style

```
pages/
  home/
    index.html
    index.scss
  about/
    about.html
    about.scss
...
```

the entry option with `names` of entrypoints (index, about, etc.):
```
entry: {
        index: 'src/views/pages/home/index.html',
        about: 'src/views/pages/about/about.html',
},
```


the `contentFunction` option in the config of the `postcss-purgecss`:
```js
   contentFunction: (sourceInputFileName) => {
      const { name } = path.parse(sourceInputFileName);

      // works after the second build, because at the first time the dist dir is empty
      return [`${paths.output}/${name}.html`];
    },
```

**Comment by vralle** [comment]
@webdiscus thanks! I tried. The directory structure solves almost all style cleanup issues other than using uncompiled templates that include parts. You helped a lot. Today was a good day. For the most part, style cleanup issues have been resolved. Also, the search for a solution led me to the discovery of UnoCss. I even managed to fall a little in love with this library, which will replace, if not Bootstrap, then probably Tailwind in my projects.

Can we close this discussion as resolved?

**Comment by vralle** [comment]
After a while, having to compile projects twice became annoying. I realized that this couldn't continue and took some measures.
The latest project looks like this:
Dir structure
![image](https://github.com/webdiscus/html-bundler-webpack-plugin/assets/3823060/e27d57da-ed29-4e3b-9953-fbd5d8959338)

`package.json`
```json
{
  ...
  "scripts": {
    "build": "npm-run-all --sequential webpack-build purgeCss html-lint",
    "purgeCss": "node config/purgeStyles.mjs",
    ...
  },
  "devDependencies": {
    "npm-run-all2": "latest",
    ...
  }
  ...
}
```
`purgeStyles.mjs:`
```js
/**
 * @typedef {import('purgecss').ResultPurge} ResultPurge
 */

import { parse, relative } from 'node:path';

import { Glob, glob } from 'glob';
import { stat, writeFile } from 'node:fs/promises';
import { PurgeCSS } from 'purgecss';

/** Project paths */
const root = new URL('../', import.meta.url).pathname;
const project = {
  output: `${root}/public`,
  outputCssDir: `${root}/public/static/css`,
};

const purgeCssShareConfig = {
  fontFace: true,
  keyframes: true,
  variables: true,
  safelist: {
    standard: ['show', 'active'],
  },
};

/** Process report */
const report = [];

/**
 * Create Glob cache for multiple uses
 * @see https://github.com/isaacs/node-glob?tab=readme-ov-file#options
 */
const cssGlobCache = new Glob(`${project.outputCssDir}/*.css`.replace(/\\/g, '/'), { nodir: true });

/** Compiled HTML files */
const htmlFiles = await glob(`${project.output}/*.html`.replace(/\\/g, '/'), { nodir: true });

/**
 * @param {string[]} cssFiles
 * @param {string} htmlFile
 */
async function addToReport(cssFiles, htmlFile) {
  const cssFilesReport = cssFiles.map(async (cssFile) => ({
    file: cssFile,
    size: (await stat(cssFile)).size,
    content: htmlFile,
  }));
  report.push(cssFilesReport);
}

/**
 * @param {string} htmlFile
 * @return {Promise<ResultPurge[]>}
 */
async function purgeContent(htmlFile) {
  const { name } = parse(htmlFile);
  const cssFiles = await glob(`${project.outputCssDir.replace(/\\/g, '/')}/${name}*.css`, { ...cssGlobCache });

  await addToReport(cssFiles, htmlFile);

  return new PurgeCSS().purge({
    ...purgeCssShareConfig,
    content: [htmlFile],
    css: cssFiles,
  });
}

async function getPurges() {
  return htmlFiles.map(async (htmlFile) => purgeContent(htmlFile));
}

/** @param {ResultPurge[][]} purges */
async function savePurges(purges) {
  return purges.map(async (resultPurge) => writeFile(resultPurge[0].file, resultPurge[0].css));
}

async function purge() {
  const purges = await getPurges();
  return savePurges(await Promise.all(purges));
}

/**
 * @param {number} size
 * @returns {number}
 */
function toKb(size) {
  return Math.round((size / 1024) * 100) / 100;
}

(async () => {
  try {
    console.info('Purge Css files...');
    const timeStart = performance.now();
    const start = await purge();
    await Promise.all(start);
    const timeEnd = performance.now();
    console.log(`Done! It took ${Math.round(timeEnd - timeStart)} ms.`);
    const table = (await Promise.all(report.flat())).map(async (cssReport) => {
      const sizeOut = (await stat(cssReport.file)).size;
      const percentage = ((cssReport.size - sizeOut) * 100) / cssReport.size;
      return {
        Styles: relative(root, cssReport.file),
        'Size before, kB': toKb(cssReport.size),
        'Size after, kB': toKb(sizeOut),
        'Purged, %': Math.round(percentage * 100) / 100,
        Content: relative(root, cssReport.content),
      };
    });
    console.table(await Promise.all(table));
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
})();
```
I use Node v20, so I wrote the script without regard to compatibility. I'm not a big JS/TS expert, but it works well.

I've achieved my goal.
![image](https://github.com/webdiscus/html-bundler-webpack-plugin/assets/3823060/3f654eb1-c26c-4a73-8bb9-d6657e64ffba)


I believe this solution is easy to integrate into any project where names of css-files contain the endpoint name.
There is also an obvious disadvantage: this solution won't be able to handle inline styles. In an ideal solution, we need a hook that provides access to the compiled templates, where we can replace the assets content.

**Reply by webdiscus** [reply]
since v3.0, the plugin has [Hooks & Callbacks](https://github.com/webdiscus/html-bundler-webpack-plugin/discussions/31#plugin-hooks-and-callbacks).
You can make your own plugin for the plugin ;-)

As an example you can see this [plugins](https://github.com/webdiscus/html-bundler-webpack-plugin/tree/master/plugins) or many test cases `hook-*`, e.g. [hook-beforeEmit](https://github.com/webdiscus/html-bundler-webpack-plugin/tree/master/test/cases/hook-beforeEmit).

**Comment by vralle** [comment]
I looked a little into the hooks. Let me summarize.

https://gist.github.com/vralle/d394e32a0f9d30798e60b3670c654fd1

**Comment by vralle** [comment]
@webdiscus, I'm trying improve my code, but think I can't do without your help.

1. I want to get a dependency tree  (all entrypoints and nested styles), something like this:
- home.html
    -- public/css/app.css
    -- public/css/home.css
- about.html
    -- public/css/app.css
    -- public/css/about.css
Can I get such a tree inside [processAssets](https://webpack.js.org/api/compilation-hooks/#processassets) hook or before?
2. Inside the hook I'm considering a strategy where one style file is used for all endpoints, the optimizer should create a new file for each endpoint and insert it into the endpoint instead of the remote one. Is it possible?
3. I've recently been cleaning up my Webpack configuration. Here's what seemed a little inconsistent:
`webpackConfig.output,filename` - the value affects all JS file names. It works globally. html-bundler-webpack-plugin can overrides file  names with `htmlBundlerPluginConfig.js.filename`. It's Ok
`webpackConfig.output,cssFilename` - doesn't have affect on CSS file names. Only `htmlBundlerPluginConfig.css.filename`. This is not critical, but when the configuration becomes complex, you want to use global settings as much as possible. This makes code maintenance a little easier

I also think that you will be a little angry after reading these questions)

**Reply by webdiscus** [reply]
hi @vralle,

why I should be angry? You can ask me anything :-)

1. It works in `afterEmit` callback/hook:
```js
new HtmlBundlerPlugin({
      verbose: true, // <= display assets tree
      entry: {
        index: 'src/views/home.html',
        about: 'src/views/about.html',
      },
      beforeEmit: (content, entry, compilation) => {
        //console.log(compilation.assets); // <= here all processed assets (content and output filename) BUT w/o integrity (if used)
        // the compilation contains only output assets filenames w/o relation to their issuer (parent)
  
       console.dir(entry, { depth: 5 }); // <= here is the assets tree of all processed resources in the Entrypoint
       // here will be available only current Entrypoint (with dependencies), not all entipoints
      },
      afterEmit: (entries) => {
        console.dir(entries, { depth: 6 }); // <= output complete assets tree exact what you need (look it self here)
        // at this stage is too late... you should self read file from disk > modify > save
      },
    }),
```

Try to use the `beforeEmit` callback/hook. This is called in the `processAssets` webpack hook as lastest (stage is Infinity), after all other plugins.

2. I think it would be possible, see the point 1.

3. I know, what webpack has `webpackConfig.output.cssFilename` (it's reserved forWebpack 6). 
You are right, the `webpackConfig.output.cssFilename` option should works like `webpackConfig.output.filename`.
I will implement it in one of next versions.
Thank you for the suggestion! Can you please create separate issue for this feature request (it's not a bug, it is not realised feature).
